crates/noirc_evaluator/src/ssa/conditional.rs

```diff
@@ -0,0 +1,387 @@
+use num_bigint::BigUint;
+use num_traits::One;
+
+use crate::ssa::node::ObjectType;
+
+use super::{
+    block::{self, BlockId, BlockType},
+    context::SsaContext,
+    inline::StackFrame,
+    node::{self, BinaryOp, Instruction, NodeId, Operation},
+};
+
+#[derive(Debug, Clone)]
+pub struct Assumption {
```

Comment:
I'm not familiar with a notion of Assumptions in Ssa, did you add it from an existing scheme/paper or does it solve a noir-specific problem? Either way, can you elaborate on the design choices here?

Reply:
Yes, it solves a noir-specific problem. Since we do not know the value of the condition of IF statements in general, we need to retrieve it and pass it along down to the generation of the constraints.
Because nested IF depends on previous conditions, I decided to represent the conditions using a decision tree, so that a node in the tree, called assumption, is a faithful and compact representation of them. It can be easily computed from its parent when we enter an IF and we simply go up in the tree when we exit the IF.
Another benefit is when we flatten the control flow, as it is much simpler to start from inner most IF up to the outer most, because in that case  we never have to handle IF containing an IF. The flattening can use the tree structure to process from the leafs to the root, without needed to parse again the CFG or to store additional data.